File: api/src/routes/roadmap.ts
import { Router } from "express";
import prisma from "../lib/prisma.js";

const router = Router();

router.get("/", async (req, res) => {
	const user = res.locals.user;
	if (!user) return res.status(401).json({ error: "Unauthorized" });

	let studentId: number | null = null;
	
	if (user.role.toUpperCase() === "STUDENT") {
		const student = await prisma.student.findUnique({ where: { userId: user.id } });
		studentId = student?.id || null;
	} else if ((user.role.toUpperCase() === "ADMIN" || user.role.toUpperCase() === "COUNSELOR") && req.query.studentId) {
		studentId = parseInt(req.query.studentId as string);
	}

	if (!studentId) return res.status(404).json({ error: "Student not found or studentId missing" });

	const student = await prisma.student.findUnique({
		where: { id: studentId },
		include: { academics: true }
	});

	if (!student) return res.status(404).json({ error: "Student not found" });

	const recommendations = [];

	// Rule: If Student.majorInterest === 'CS' AND Student.grade === 10 -> Recommend 'AP CS A' for 11th grade
	if (student.majorInterest === "CS" && student.grade === 10) {
		recommendations.push({
			year: 11,
			course: "AP Computer Science A",
			reason: "To strengthen your CS foundation for college applications."
		});
	}

	// Simple vertical list of years/courses
	const roadmap = [
		{ year: 9, courses: ["Algebra II Honors", "English 9 Honors", "Biology Honors", "Spanish I", "World History"] },
		{ year: 10, courses: ["Pre-Calculus Honors", "English 10 Honors", "Chemistry Honors", "Spanish II", "AP World History"] },
		{ year: 11, courses: ["AP Calculus AB", "AP Physics 1", "English 11 AP Lang", "Spanish III", "AP US History"] },
		{ year: 12, courses: ["AP Calculus BC", "AP Physics C", "English 12 AP Lit", "Spanish IV Honors", "AP Gov/Econ"] }
	];

	// Merge recommendations
	recommendations.forEach(rec => {
		const year = roadmap.find(r => r.year === rec.year);
		if (year) {
			year.courses.push(`* ${rec.course} (Recommended)`);
		}
	});

	res.json(roadmap);
});

export default router;

File: api_py/routers/roadmap.py
from fastapi import APIRouter, Depends, HTTPException, Request
from ..database import execute_query
from ..auth import get_current_user
from typing import Optional
import traceback

router = APIRouter()

@router.get("/")
async def get_roadmap(studentId: Optional[str] = None, current_user: dict = Depends(get_current_user)):
    try:
        user_id = current_user["id"]
        user_role = current_user["role"].upper()
        print(f"[ROADMAP] User: {user_id}, Role: {user_role}, Requested studentId: {studentId}")
        
        target_student_id = None
        major_interest = "Other"
        current_grade = 9
        
        if user_role == "STUDENT":
            student = execute_query("SELECT id, grade, majorInterest FROM Student WHERE userId = %s", (user_id,), fetch_one=True)
            if student:
                target_student_id = student["id"]
                current_grade = student["grade"]
                major_interest = student["majorInterest"]
        elif (user_role in ["ADMIN", "COUNSELOR"]) and studentId:
            try:
                target_student_id = int(studentId)
                student = execute_query("SELECT grade, majorInterest FROM Student WHERE id = %s", (target_student_id,), fetch_one=True)
                if student:
                    current_grade = student["grade"]
                    major_interest = student["majorInterest"]
                else:
                    print(f"[ROADMAP] Student with ID {target_student_id} not found in database.")
                    raise HTTPException(status_code=404, detail=f"Student with ID {target_student_id} not found")
            except ValueError:
                raise HTTPException(status_code=400, detail="Invalid student ID format")
                
        if not target_student_id:
            if user_role in ["ADMIN", "COUNSELOR"]:
                 return [] # Return empty for staff if no student selected yet
            print(f"[ROADMAP] No target_student_id resolved for user {user_id}")
            raise HTTPException(status_code=400, detail="Student ID required")
            
        recommendations = []
        if major_interest == "CS" and current_grade == 10:
            recommendations.append({
                "year": 11,
                "course": "AP Computer Science A",
                "reason": "To strengthen your CS foundation for college applications."
            })
            
        roadmap = [
            {"year": 9, "courses": ["Algebra II Honors", "English 9 Honors", "Biology Honors", "Spanish I", "World History"]},
            {"year": 10, "courses": ["Pre-Calculus Honors", "English 10 Honors", "Chemistry Honors", "Spanish II", "AP World History"]},
            {"year": 11, "courses": ["AP Calculus AB", "AP Physics 1", "English 11 AP Lang", "Spanish III", "AP US History"]},
            {"year": 12, "courses": ["AP Calculus BC", "AP Physics C", "English 12 AP Lit", "Spanish IV Honors", "AP Gov/Econ"]}
        ]
        
        for rec in recommendations:
            for r in roadmap:
                if r["year"] == rec["year"]:
                    r["courses"].append(f"* {rec['course']} (Recommended)")
                    
        return roadmap
    except HTTPException:
        raise
    except Exception as e:
        print(f"[ROADMAP] Exception in GET /: {traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))
